src/Core/Context/LegacyControllerContextBuilder.php

```diff
@@ -0,0 +1,224 @@
+use PrestaShop\PrestaShop\Core\Util\Inflector;
+use PrestaShopBundle\Entity\Repository\TabRepository;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use Tools;
+
+class LegacyControllerContextBuilder implements LegacyContextBuilderInterface
+{
+    public function __construct(
+        private readonly EmployeeContext $employeeContext,
+        private readonly ShopContext $shopContext,
+        private readonly ContextStateManager $contextStateManager,
+        private readonly RequestStack $requestStack,
+        private readonly array $controllersLockedToAllShopContext,
+        private readonly TabRepository $tabRepository,
+        private readonly TranslatorInterface $translator,
+        private readonly MultistoreFeature $multistoreFeature,
+        private readonly ContainerInterface $container,
+        private readonly string $coreDir,
+    ) {
+    }
+
+    public function buildLegacyContext(): void
+    {
+        $request = $this->requestStack->getMainRequest();
+
+        $controllerName = $this->getControllerName($request);
+        $multiShopContext = $this->getMultiShopContext($controllerName);
+        $id = $this->getTabId($controllerName);
+        $token = null;
+        if ($this->employeeContext->getEmployee()) {
+            $token = Tools::getAdminToken($controllerName . $id . $this->employeeContext->getEmployee()->getId());
+        }
+        $conf = $this->getConf();
+        $error = $this->getError();
+        $shopLinkType = $this->getShopLinkType($controllerName);
+        $overrideFolder = Tools::toUnderscoreCase(substr($controllerName, 5)) . '/';
+        $adminWebPath = $this->getAdminWebPath();
+        $controllerType = 'admin';
+        $className = $this->getClassName($controllerName);
+
+        $controllerContext = new LegacyController(
+            $this->container,
+            $controllerName,
+            $controllerType,
+            $controllerName,
+            $multiShopContext,
+            $className,
+            $id,
+            $token,
+            $conf,
+            $error,
+            $shopLinkType,
+            $overrideFolder,
+            $overrideFolder,
+            $adminWebPath
+        );
+
+        $this->contextStateManager->setController($controllerContext);
+    }
+
+    private function getControllerName(?Request $request): string
+    {
+        $controllerName = 'AdminController';
+
+        if ($request->attributes->has('_legacy_controller')) {
+            $controllerName = $request->attributes->get('_legacy_controller');
+        } elseif ($request->query->has('controller')) {
+            $controllerName = $request->query->get('controller');
+        }
+
+        return $controllerName;
+    }
+
+    private function getClassName(string $controllerName): ?string
```

Comment:
I think it's worth adding a bit of detail and comment here about the purpose of the method (getting the associated legacy object based on the controller name basically), especially in the default case that is a little obscure at first sight with a mix of regexp and inflection